### ISSUE TITLE
Provide ability to detect the debug port for a stack

### ISSUE THREAD

**Issue by makandre** [appsody/stacks#261]
**Is your feature request related to a problem? Please describe.**

I would like to be able determine the debug port for a stack, to facilitate integration with IDE tooling, such as Codewind

I can see in all stack images, the `PORT` environment variable is defined specifying the application port.  However, nothing is defined for the debug port.

In cases where the stack exposes more than 2 ports, I need to guess which is the debug port. e.g. for `java-microprofile`, it has:

````
ENV PORT=9080

EXPOSE 9080
EXPOSE 9443
EXPOSE 7777
````

Here I *know* 7777 is for debug, but in code, how does it know?

**Describe the solution you'd like**

~Is it possible to have `ENV DEBUG_PORT` defined?~

**Describe alternatives you've considered**

**Additional context**

**Comment by skoh7645** [comment]
@makandre There is a deployment yaml inside each stack -> `repo/stack/image/config/app-deploy.yaml`

**Comment by makandre** [comment]
Looking at https://github.com/appsody/stacks/blob/master/incubator/java-microprofile/image/config/app-deploy.yaml but it doesn't look like it contains anything that would tell me what the debug port is

**Comment by neeraj-laad** [comment]
@kewegner @kylegc Would like to get your thoughts on this.

**Comment by makandre** [comment]
I did some research on how to look at a file inside the docker images.  Seems that you either have to actually run it then look inside, or use `docker save` command but that saves the entire filesys into an archive.

Neither option seems ideal and would be costly in performance.  So I would still prefer to get the debug port by inspection of the image.

**Comment by kylegc** [comment]
So funny thing, we used to have an APPSODY_DEBUG_PORT stack variable and we got rid of it in favor of EXPOSE. This was done for flexibility in specifying multiple ports and consistency with built-in docker constructs. There was a scenario where Spring needed an additional debug port for livereload.

@makandre Generally you need a way for the stack to communicate what runs on each exposed port. For instance, how do you know the main application port? The env PORT is a good guess, but the stacks don't have to define this. They could leave off PORT and just use EXPOSE or set it to something else.

We've discussed having each stack document their main endpoints (health, metrics, etc) in stack.yaml. We could also document the application and debug ports there instead of cluttering the Dockerfile-stack.

Thoughts? @neeraj-laad @chilanti @ebullient

**Comment by makandre** [comment]
Unfortunately if we only had the `EXPOSE` ports to look at, Codewind then needs to guess which one is the main application port and which one is the main debug port; and we can't always guess right... hence at the moment we actually have to put the ports in a [file](https://github.com/eclipse/codewind-appsody-extension/blob/master/.sh-extension) and have some logic to figure out what ports are for which stack.

This means when a new stack is released (and it shows up automatically in Codewind because we read the latest index via the CLI), it may not work correctly until we update the ports.  This is the problem we're trying to solve with this request.

**Comment by neeraj-laad** [comment]
@makandre I agree, having to guess ports/endpoints might not be the right approach for Codewind. 

That said if stacks need multiple ports of one type (RUN/DEBUG) then using these variables will only address some stacks not all.

What we have been thinking is that we extract this information out in a descriptor file (stack.yaml or similar) and then have the information propagated to appsody artefacts (index or stack image etc.). so that tools that need this information van extract this programmatically in a consistent way.

**Comment by sghung** [comment]
@neeraj-laad any update on this issue? Will it make it by Oct 9, which is the Codewind 0.5.0 DCUT?

**Comment by makandre** [comment]
We'll need some lead time to react

**Comment by makandre** [comment]
Updating this old issue with some new info from https://github.com/eclipse/codewind/issues/282#issuecomment-576893988

This gist of it is, my original suggestion of defining an environment variable for the debug port is not going to work for us.
